### PR TITLE
Add DetectionOutputLayer and MultiBoxLossLayer.

### LINKED ISSUE #2496: Add DetectionOutputLayer and MultiBoxLossLayer for SSD
We do generation operations in DetectionOutputLayer, key operations include NMS、filtering、decoding etc.
MultiBoxLossLayer calculates losses include soomthL1 loss and classification neg-log loss.

### PR DESCRIPTION
**Pull request by pkuyym**
resolves #2496

### CODE REVIEW

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+      // Negative mining samples
+      for (size_t i = 0; i < allNegIndices_[n].size(); ++i) {
+        confGTData_->getData()[count] = backgroundId_;
+        size_t confOffset =
+            n * numPriors_ * numClasses_ + allNegIndices_[n][i] * numClasses_;
+        for (size_t j = 0; j < numClasses_; ++j) {
+          confProb_->getData()[count * numClasses_ + j] =
+              (confBuffer_->getData() + confOffset)[j];
+          confPredData.push_back((confBuffer_->getData() + confOffset)[j]);
+        }
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: for循环里的拷贝可以使用std::copy.

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        if (allMatchIndices_[n][i] == -1) continue;
+        size_t labelOffset = (labelIndex[n] + allMatchIndices_[n][i]) * 6;
+        const int gtLabel = (labelValue->getData() + labelOffset)[0];
+        confGTData_->getData()[count] = gtLabel;
+        size_t confOffset = n * numPriors_ * numClasses_ + i * numClasses_;
+        for (size_t j = 0; j < numClasses_; ++j) {
+          confProb_->getData()[count * numClasses_ + j] =
+              (confBuffer_->getData() + confOffset)[j];
+          confPredData.push_back((confBuffer_->getData() + confOffset)[j]);
+        }
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: `for`循环里的连续拷贝可以使用`std::copy`.

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+    for (size_t n = 0; n < batchSize; ++n) {
+      for (size_t i = 0; i < numPriors_; ++i) {
+        if (allMatchIndices_[n][i] == -1) continue;  // match none
+        size_t locOffset =
+            n * (locBuffer_->getElementCnt() / batchSize) + i * 4;
+        locDiff_->getData()[count++] = (locBuffer_->getData() + locOffset)[0];
+        locDiff_->getData()[count++] = (locBuffer_->getData() + locOffset)[1];
+        locDiff_->getData()[count++] = (locBuffer_->getData() + locOffset)[2];
+        locDiff_->getData()[count++] = (locBuffer_->getData() + locOffset)[3];
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: `locDiff_->getData()`, `locBuffer_->getData()`，放到`for`循环外面，里面可以对移动指针。 连续拷贝可以用`std::copy`或者`std::memcpy`吧

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    // Copy gradient back
+    size_t count = 0;
+    for (size_t n = 0; n < batchSize; ++n)
+      for (size_t i = 0; i < numPriors_; ++i) {
+        if (allMatchIndices_[n][i] == -1) continue;
+        real* locDiffData = locBuffer_->getData() + n * numPriors_ * 4 + i * 4;
+        locDiffData[0] = (locDiff_->getData() + count * 4)[0];
+        locDiffData[1] = (locDiff_->getData() + count * 4)[1];
+        locDiffData[2] = (locDiff_->getData() + count * 4)[2];
+        locDiffData[3] = (locDiff_->getData() + count * 4)[3];
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: 同样，`locBuffer_->getData()`, `locDiff_->getData()`，里面可以对指针移动~

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        ++count;
+      }
+    CHECK_EQ(count, numMatches_);
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: forward也check下吧~

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        for (size_t j = 0; j < numClasses_; ++j)
+          confDiffData[j] = (confProb_->getData() + count * numClasses_)[j];
+        ++count;
+      }
+      for (size_t i = 0; i < allNegIndices_[n].size(); ++i) {
+        int idx = allNegIndices_[n][i];
+        real* confDiffData = confBuffer_->getData() +
+                             n * numPriors_ * numClasses_ + idx * numClasses_;
+        for (size_t j = 0; j < numClasses_; ++j)
+          confDiffData[j] = (confProb_->getData() + count * numClasses_)[j];
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: 这块两处拷贝同上，可以去掉for循环，使用std::copy or std::memcpy.

File: python/paddle/trainer/config_parser.py [hunk 7]
```diff
@@ -1676,6 +1676,52 @@ def __init__(self, name, inputs, size, min_size, max_size, aspect_ratio,
         self.config.size = size
 
 
+@config_layer('multibox_loss')
+class MultiBoxLossLayer(LayerBase):
+    def __init__(self, name, inputs, input_num, num_classes, overlap_threshold,
+                 neg_pos_ratio, neg_overlap, background_id):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: `background_id):` ->  `background_id,  **xargs):`

File: python/paddle/trainer/config_parser.py [hunk 8]
```diff
@@ -1676,6 +1676,52 @@ def __init__(self, name, inputs, size, min_size, max_size, aspect_ratio,
...
+        self.config.inputs[0].multibox_loss_conf.background_id = background_id
+        self.config.inputs[0].multibox_loss_conf.input_num = input_num
+        self.config.size = 1
+
+
+@config_layer('detection_output')
+class DetectionOutputLayer(LayerBase):
+    def __init__(self, name, inputs, size, input_num, num_classes,
+                 nms_threshold, nms_top_k, keep_top_k, confidence_threshold,
+                 background_id):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: background_id): -> background_id, **xargs):

File: python/paddle/trainer_config_helpers/layers.py [hunk 9]
```diff
@@ -1052,6 +1056,163 @@ def priorbox_layer(input,
...
+    """
+    input_loc_num = 0
+    input_conf_num = 0
+
+    if isinstance(input_loc, LayerOutput):
+        input_loc = [input_loc]
+    assert isinstance(input_loc, collections.Sequence)  # list or tuple
+    for each in input_loc:
+        assert isinstance(each, LayerOutput)
+        input_loc_num += 1
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 可以去掉前面input_loc_num=0，改成：
```
for each in input_loc:
    assert isinstance(each, LayerOutput)
input_loc_num = len(input_loc)
```

input_conf_num同样~

File: python/paddle/trainer_config_helpers/layers.py [hunk 10]
```diff
@@ -1052,6 +1056,163 @@ def priorbox_layer(input,
...
+
+    if isinstance(input_conf, LayerOutput):
+        input_conf = [input_conf]
+    assert isinstance(input_conf, collections.Sequence)  # list or tuple
+    for each in input_conf:
+        assert isinstance(each, LayerOutput)
+        input_conf_num += 1
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 上同~

File: python/paddle/trainer_config_helpers/layers.py [hunk 11]
```diff
@@ -115,6 +115,8 @@
     'print_layer',
     'priorbox_layer',
     'cross_channel_norm_layer',
+    'multibox_loss_layer',
+    'detection_output_layer',
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 增加到文档里，同时增加单测~

Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: `doc/api/v2/config/layer.rst`

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+    confProb_->softmax(*confProb_);
+    MatrixPtr confLossOutput;
+    Matrix::resizeOrCreate(confLossOutput, numConf_, 1, false, false);
+    confLossOutput->oneHotCrossEntropy(*confProb_, *confGTData_);
+    confLoss_ = confLossOutput->getSum() / numMatches_;
+  }
+  real loss = locLoss_ + confLoss_;
+  MatrixPtr outV = getOutputValue();
+  std::vector<real> tmp(batchSize, loss);
+  outV->copyFrom(&tmp[0], batchSize);
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: 每个样本的loss都一样？如果是，直接赋值给outV就可以，不用中间构造:
```
std::vector<real> tmp(batchSize, loss);
```
看下 `BaseMatrix::assign`是否可用。

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+  size_t locOffset = 0;
+  size_t confOffset = 0;
+  auto& layerConf = config_.inputs(0).multibox_loss_conf();
+  for (size_t n = 0; n < inputNum_; ++n) {
+    const MatrixPtr inLoc = getInputValue(*getLocInputLayer(n));
+    const MatrixPtr inConf = getInputValue(*getConfInputLayer(n));
+    size_t height = getInput(*getLocInputLayer(n)).getFrameHeight();
+    if (!height) height = layerConf.height();
+    size_t width = getInput(*getLocInputLayer(n)).getFrameWidth();
+    if (!width) width = layerConf.width();
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]: `layerConf.height()`和`layerConf.width()`为了单测才加的，仅仅在这里注释下。

File: paddle/gserver/layers/MultiBoxLossLayer.cpp [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+  size_t locOffset = 0;
+  size_t confOffset = 0;
+  auto layerConf = config_.inputs(0).multibox_loss_conf();
+  for (size_t n = 0; n < inputNum_; ++n) {
+    const MatrixPtr inLocG = getInputGrad(*getLocInputLayer(n));
+    const MatrixPtr inConfG = getInputGrad(*getConfInputLayer(n));
+    size_t height = getInput(*getLocInputLayer(n)).getFrameHeight();
+    if (!height) height = layerConf.height();
+    size_t width = getInput(*getLocInputLayer(n)).getFrameWidth();
+    if (!width) width = layerConf.width();
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.cpp]:
```
size_t width = getInput(*getLocInputLayer(n)).getFrameWidth();
```
这句理解，每个LocInputLayer的height/width可以不一样，实际也不一样吧? 但为了单测加的：

```
if (!width) width = layerConf.width();
```
确实每个LocInputLayer的height/width都一样。

File: paddle/gserver/layers/MultiBoxLossLayer.h [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+using std::pair;
+
+namespace paddle {
+
+/**
+ * The multibox loss layer for a SSD detection task.
+ * The loss is composed by the location loss and the confidence loss.
+ * The location loss is a smooth L1 loss and the confidence loss is
+ * a softmax loss.
+ * - Input: This layer need four input layers: This first input layer
```
Comment (qingqing01) [inline paddle/gserver/layers/MultiBoxLossLayer.h]: need -> needs
This first input -> The first input

File: paddle/gserver/layers/DetectionOutputLayer.h [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+#include <map>
+#include <vector>
+#include "DetectionUtil.h"
+#include "Layer.h"
+
+namespace paddle {
+
+/**
+ * The detection output layer for a SSD detection task. This layer apply the
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionOutputLayer.h]: apply -> applies

File: paddle/gserver/layers/DetectionOutputLayer.h [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+ * Non-maximum suppression to the all predicted bounding box and keep the
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionOutputLayer.h]: keep -> keeps

File: paddle/gserver/layers/DetectionOutputLayer.h [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+ * Top-K bounding boxes.
+ * - Input: This layer needs three input layers: This first input layer
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionOutputLayer.h]: This first input layer -> The first input layer

File: paddle/gserver/layers/DetectionOutputLayer.h [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+ *          is the priorbox layer. The rest two input layers are convolution
+ *          layers for generating bbox location offset and the classification
+ *          confidence.
+ * - Output: The predict bounding box location.
```
Comment (qingqing01) [inline paddle/gserver/layers/DetectionOutputLayer.h]: The predicted bounding box locations.

File: doc/api/v2/config/layer.rst [hunk 20]
```diff
@@ -468,3 +473,11 @@ prelu
 --------
 ..  autoclass:: paddle.v2.layer.prelu
     :noindex:
+
+Detection output Layer
+======================
+
+detection_output
+---
```
Comment (qingqing01) [inline doc/api/v2/config/layer.rst]: ---加长

File: python/paddle/trainer_config_helpers/layers.py [hunk 21]
```diff
@@ -1052,6 +1056,160 @@ def priorbox_layer(input,
...
+                        neg_pos_ratio=3.0,
+                        neg_overlap=0.5,
+                        background_id=0,
+                        name=None):
+    """
+    Compute the location loss and the confidence loss for ssd.
+
+    :param name: The Layer Name.
+    :type name: basestring
+    :param input_loc: The input predict location.
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: The input layer of predicted locations.

File: python/paddle/trainer_config_helpers/layers.py [hunk 22]
```diff
@@ -1052,6 +1056,160 @@ def priorbox_layer(input,
...
+    :type input_loc: LayerOutput
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: LayerOutput | List of LayerOutput

File: python/paddle/trainer_config_helpers/layers.py [hunk 23]
```diff
@@ -1052,6 +1056,160 @@ def priorbox_layer(input,
...
+    :param input_conf: The input priorbox confidence.
+    :type input_conf: LayerOutput
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: LayerOutput | List of LayerOutput

File: python/paddle/trainer_config_helpers/layers.py [hunk 24]
```diff
@@ -1052,6 +1056,160 @@ def priorbox_layer(input,
...
+                           confidence_threshold=0.01,
+                           background_id=0,
+                           name=None):
+    """
+    Apply the NMS to the output of network and compute the predict bounding
+    box location.
+
+    :param name: The Layer Name.
+    :type name: basestring
+    :param input_loc: The input predict location.
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: The input layer of  predicted locations

File: python/paddle/trainer_config_helpers/layers.py [hunk 25]
```diff
@@ -1052,6 +1056,160 @@ def priorbox_layer(input,
...
+    :type input_loc: LayerOutput
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: LayerOutput | List of LayerOutput.

File: python/paddle/trainer_config_helpers/layers.py [hunk 26]
```diff
@@ -1052,6 +1056,160 @@ def priorbox_layer(input,
...
+    :param input_conf: The input priorbox confidence.
+    :type input_conf: LayerOutput
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: LayerOutput | List of LayerOutput.

File: python/paddle/trainer_config_helpers/layers.py (hunk 9 above)
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 这块没有Done.

File: python/paddle/trainer_config_helpers/layers.py (hunk 10 above)
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 这块也没有Done.

File: python/paddle/trainer_config_helpers/tests/configs/test_multibox_loss_layer.py [hunk 27]
```diff
@@ -0,0 +1,25 @@
+from paddle.trainer_config_helpers import *
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/tests/configs/test_multibox_loss_layer.py]: 需要运行下单测，将protostr文件add进来。上同。
